File: android-sample-java/app/src/main/java/io/logto/demo4j/fragment/LoginFragment.java
package io.logto.demo4j.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import io.logto.demo4j.R;
import io.logto.demo4j.viewmodel.LogtoViewModel;
import io.logto.demo4j.viewmodel.LogtoViewModelFactory;

public class LoginFragment extends Fragment {

    private LogtoViewModel logtoViewModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_login, container, false);
        initView(view);
        initViewModel(view);
        return view;
    }

    private void initView(View view) {
        view.findViewById(R.id.sign_in_button).setOnClickListener( v -> {
            logtoViewModel.signIn(requireActivity());
        });
    }

    private void initViewModel(View view) {
        LogtoViewModelFactory logtoViewModelFactory = new LogtoViewModelFactory(requireActivity().getApplication());
        logtoViewModel = new ViewModelProvider(requireActivity(), logtoViewModelFactory).get(LogtoViewModel.class);
        logtoViewModel.isAuthenticated().observe(getViewLifecycleOwner(),  hasAuthenticated -> {
            if (hasAuthenticated) {
                Navigation.findNavController(view).navigate(R.id.action_loginFragment_to_tokenFragment);
            }
        });
    }
}

File: android-sample-kotlin/app/src/main/kotlin/io/logto/android/sample4k/fragment/LoginFragment.kt
package io.logto.android.sample4k.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import io.logto.android.sample4k.R
import io.logto.android.sample4k.viewmodel.LogtoViewModel
import io.logto.android.sample4k.viewmodel.LogtoViewModelFactory

class LoginFragment : Fragment() {

    private val logtoViewModel: LogtoViewModel by activityViewModels {
        LogtoViewModelFactory(requireActivity().application)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        return inflater.inflate(R.layout.fragment_login, container, false).also {
            initView(it)
            initViewModel()
        }
    }

    private fun initView(view: View) {
        view.findViewById<Button>(R.id.sign_in_button).setOnClickListener {
            logtoViewModel.signIn(requireActivity())
        }
    }

    private fun initViewModel() {
        logtoViewModel.authenticated.observe(viewLifecycleOwner) { hasAuthenticated ->
            if (hasAuthenticated) {
                findNavController().navigate(R.id.action_loginFragment_to_tokenFragment)
            }
        }
    }
}
